File: frontend/src/utils/shipmentUtils.ts
// Status có thể hủy Shipment của Manager
export const canCancelManagerShipment = (value: string) => {
  return ["PENDING"].includes(value)
};

// Status có thể edit Shipment của Manager
export const canEditManagerShipment = (value: string) => {
  return ["PENDING"].includes(value)
};

// Status có thể change orders in Shipment của Manager
export const canEditOrdersManagerShipment = (value: string, type: string) => {
  return (["PENDING", "IN_TRANSIT"].includes(value) && type === "DELIVERY") ||
      (["PENDING"].includes(value) && type === "TRANSFER");
};

export const canConfirmDestinationOrdersManagerShipment = (value: string) => {
  return ["COMPLETED"].includes(value);
};

export const canDeleteOrdersManagerShipment = (value: string) => {
  return ["PENDING"].includes(value)
};

export const SHIPMENT_STATUSES = ['PENDING', 'IN_TRANSIT', 'COMPLETED', 'CANCELLED'] as const;
export const translateShipmentStatus = (value: string): string => {
  switch (value) {
    case 'PENDING':
      return 'Chờ khởi hành';
    case 'IN_TRANSIT':
      return 'Đang vận chuyển';
    case 'COMPLETED':
      return 'Đã hoàn thành';
    case 'CANCELLED':
      return 'Đã hủy';
    default:
      return value;
  }
};

export const SHIPMENT_TYPES = ['DELIVERY', 'TRANSFER'] as const;
export const translateShipmentType = (value: string): string => {
  switch (value) {
    case 'DELIVERY':
      return 'Giao hàng';
    case 'TRANSFER':
      return 'Trung chuyển';
    default:
      return value;
  }
};

export const SHIPMENT_FILTER_SORT = [
  'NEWEST',
  'OLDEST',
] as const;
export const translateShipmentFilterSort = (value: string): string => {
  switch (value) {
    case 'NEWEST': return 'Mới nhất';
    case 'OLDEST': return 'Cũ nhất';
    default:
      return value;
  }
};

File: backend/src/main/java/com/logistics/utils/ShipmentUtils.java
package com.logistics.utils;

import com.logistics.enums.ShipmentStatus;
import com.logistics.enums.ShipmentType;

import java.util.Set;

public class ShipmentUtils {

    // Những Status mà Manager được phép hủy
    private static final Set<ShipmentStatus> CANCELLABLE_STATUSES_BY_MANAGER = Set.of(
            ShipmentStatus.PENDING);

    public static boolean canManagerCancelShipment(ShipmentStatus status) {
        return CANCELLABLE_STATUSES_BY_MANAGER.contains(status);
    }

    // Những Status mà Manager được phép chỉnh sửa
    private static final Set<ShipmentStatus> EDITABLE_STATUSES_BY_MANAGER = Set.of(
            ShipmentStatus.PENDING);

    public static boolean canManagerEditShipment(ShipmentStatus status) {
        return EDITABLE_STATUSES_BY_MANAGER.contains(status);
    }

    // Những Status mà Manager được phép thêm đơn hàng vào chuyến
    private static final Set<ShipmentStatus> ADDED_ORDER_STATUSES_BY_MANAGER = Set.of(
            ShipmentStatus.PENDING);

    public static boolean canManagerAddOrderForShipment(ShipmentStatus status) {
        return ADDED_ORDER_STATUSES_BY_MANAGER.contains(status);
    }

    public static String translateShipmentStatus(ShipmentStatus value) {
        if (value == null) return "";

        return switch (value) {
            case PENDING -> "Chờ khởi hành";
            case IN_TRANSIT -> "Đang vận chuyển";
            case COMPLETED -> "Đã hoàn thành";
            case CANCELLED -> "Đã hủy";
            default -> value.name();
        };
    }

    public static String translateShipmentType(ShipmentType value) {
        if (value == null) return "";

        return switch (value) {
            case DELIVERY -> "Giao hàng";
            case TRANSFER -> "Trung chuyển";
            default -> value.name();
        };
    }
}